File: system_simulations/network_model/java/discrete_event/Machine.java
import java.util.ArrayList;
import java.math.BigDecimal;

public class Machine<Input, Output> implements AtomicModel<Input, Output> {
	
	private static Scheduler<Machine> scheduler;
	private static Time currentTime = new Time(new BigDecimal(0.0), 0);
	private static int count = 0;
	private BigDecimal s;
	private BigDecimal t;
	private ArrayList<Input> inputs;
	private ArrayList<Output> outputs;
	private ArrayList<Event> eventLog;
	private String name;
	private Time lastTime;
	private int p;

	public Machine(BigDecimal t, String name){
		this.name = name;
		this.t = t;
		this.p = 0;
		this.s = new BigDecimal(0.0);
		this.inputs = new ArrayList<Input>();
		this.outputs = new ArrayList<Output>();
		this.eventLog = new ArrayList<Event>();
		this.lastTime = new Time(new BigDecimal(0.0), 0);
	}

	public int lambda(){
		return 1;
	}

	public void deltaInternal(){
		lastTime = currentTime;
		p--;
		s = t;
		
		for (Output output : outputs){
			if (Network.class.isInstance(output)){
				System.out.println("Network output: " + (++count));
			} else {
				Event<Machine> event = Event.builder(new Time(currentTime.getReal(), 0), "deltaExternal", output).addParameter(lambda()).build();	
				scheduler.put(event);	
			}
		}

		if (p > 0){
			Event<Machine> event = Event.builder(new Time(currentTime.getReal().add(timeAdvance()), 0), "deltaInternal", this).build();	
			scheduler.put(event);
		}
	}

	public void deltaExternal(BigDecimal e, int q){
		lastTime = currentTime;
		
		Event<Machine> event = null;
		if (p > 0){
			p += q;
			s = s.subtract(e);

			if(scheduler.remove(scheduler.find("deltaInternal", this))){
				event = Event.builder(new Time(currentTime.getReal().add(timeAdvance()), 0), "deltaInternal", this).build();
			}

		} else {
			p += q;
			s = t;
			event = Event.builder(new Time(currentTime.getReal().add(t), 0), "deltaInternal", this).build();	
		}
		scheduler.put(event);
	}

	public void deltaConfluent(int q){
		lastTime = currentTime;
		p += (q - 1);
		s = t;

		Event<Machine> outputEvent = Event.builder(new Time(currentTime.getReal().add(timeAdvance()), 0), "deltaInternal", this).build();
		scheduler.put(outputEvent);

		for (Output output : outputs){
			if (Network.class.isInstance(output)){
				System.out.println("Network output: " + (++count));		
			} else {
				Event<Machine> inputEvent = Event.builder(new Time(currentTime.getReal(), 0), "deltaExternal", output).addParameter(new BigDecimal("0.0")).addParameter(lambda()).build();
				scheduler.put(inputEvent);		
			}	
		}
	}

	public BigDecimal timeAdvance(){
		if (p > 0){
			return s;
		} else {
			return new BigDecimal(Integer.MAX_VALUE);
		}
	}

	public Time getLast(){
		return lastTime;
	}

	public void log(Event e){
		eventLog.add(e);
	}

	public void addInput(Input I){
		inputs.add(I);
	}

	public void addOutput(Output O){
		outputs.add(O);
	}

	public static void setTime(Time t){
		currentTime = t;
	}

	public static Time getTime(){
		return currentTime;
	}

	public static BigDecimal getRealTime(){
		return currentTime.getReal();
	}

	public static void setScheduler(Scheduler s){
		scheduler = s;
	}

	public static Scheduler getScheduler(){
		return scheduler;
	}

	public String name(){
		return name;
	}

	public String toString(){
		return name + ": | p: " + p + " | s: " + s;
	}
}

File: system_simulations/network_model/cpp/discrete_event/Machine.cpp
#ifndef MACHINE
#define MACHINE
#include <string>
#include <vector>
#include <typeinfo>
#include <sstream>
#include "Network.cpp"
#include "AtomicModel.cpp"
#include "Time.cpp"
#include "Scheduler.cpp"
#include "Event.cpp"
using namespace std;

template<class I, class O>
class Machine : public AtomicModel {
	public:
		//Time* last_time;

		Machine(double t, string name, Scheduler<AtomicModel>* scheduler){
			this->t = t;
			this->name = name;
			this->s = 0;
			this->p = 0;
			this->last_time = new Time(0.0, 0);
			this->scheduler = scheduler;
			//inputs.push_back(input);
			//outputs.push_back(output);
		}

		int lambda(){
			return 1;
		}

		void delta_internal(){
			last_time = current_time;
			p--;
			s = t;
			for (O* output : outputs){
				if (typeid(O) == typeid(Network<AtomicModel>)){
					cout << "Network output: " << (++count) << endl;
				} else {
					double x = current_time->get_real();
					Event<O>* event = new Event<O>(new Time(x, 0), EXTERNAL, output, lambda());
					scheduler->put((Event<AtomicModel>*) event);
				}
			}

			if (p > 0){
				double x = current_time->get_real() + time_advance();
				Event<AtomicModel>* event = new Event<AtomicModel>(new Time(x,0), INTERNAL, this);
				scheduler->put(event);
			}
		}

		void delta_external(double e, int q){
			last_time = current_time;
			
			if (p > 0){
				p += q;
				s -= e;
				if (scheduler->remove(scheduler->find(INTERNAL, this))){
					double x = current_time->get_real() + time_advance();
					Event<AtomicModel>* new_event = new Event<AtomicModel>(new Time(x, 0), INTERNAL, this);
					scheduler->put(new_event);
				}
			} else {
				p += q;
				s = t;

				double x = current_time->get_real() + t;
				Event<AtomicModel>* event = new Event<AtomicModel>(new Time(x, 0), INTERNAL, this);
				scheduler->put(event);
			}
		}

		void delta_confluent(int q){
			last_time = current_time;
			p += (q - 1);
			s = t;

			double x = current_time->get_real() + time_advance();
			Event<AtomicModel>* output_event = new Event<AtomicModel>(new Time(x,0), INTERNAL, this);
			//cout << "EVENT: "; output_event->print();
			scheduler->put(output_event);

			for (O* output : outputs){
				if (typeid(O) == typeid(Network<AtomicModel>)){
					cout << "Network output: " << (++count) << endl;
				} else {
					double x = current_time->get_real();
					Event<O>* event = new Event<O>(new Time(x, 0), EXTERNAL, output, lambda());
					//cout << "EVENT2: "; event->print();
					scheduler->put((Event<AtomicModel>*) event);
				}
			}
		}

		double time_advance(){
			if (p > 0){
				return s;
			} else {
				return 9999999.99999;
			}
		}

		string to_string(int i){
		    stringstream ss;
		    ss << i;
		    return ss.str();
		}

		string get_name(){
			string str = to_string(p);
			string ret_val = name + " | p: " + str;
			return ret_val;
		}

		void add_input(I* in){
			inputs.push_back(in);
		}

		void add_output(O* out){
			outputs.push_back(out);
		}

	private:
		string name;
		double t;
		double s;
		int p;
		int count = 0;
		vector<I*> inputs;
		vector<O*> outputs;
		string INTERNAL = "delta_internal";
		string EXTERNAL = "delta_external";
		Scheduler<AtomicModel>* scheduler;
};

#endif